lib/maths/CBoostedTreeImpl.cc

```diff
@@ -1453,31 +1751,38 @@ bool CBoostedTreeImpl::selectNextHyperparameters(const TMeanVarAccumulator& loss
 
 void CBoostedTreeImpl::captureBestHyperparameters(const TMeanVarAccumulator& lossMoments,
                                                   std::size_t maximumNumberTrees,
-                                                  double numberNodes) {
+                                                  double fixedNumberNodes,
+                                                  double numberExtraNodes) {
     // We capture the parameters with the lowest error at one standard
     // deviation above the mean. If the mean error improvement is marginal
     // we prefer the solution with the least variation across the folds.
 
-    // Add 0.01 * "forest number nodes" * E[GP] / "average forest number nodes" to meanLoss.
-    double modelSizeDifferentiator{0.01 * numberNodes /
-                                   CBasicStatistics::mean(m_ForestSizeAccumulator) *
-                                   CBasicStatistics::mean(m_MeanLossAccumulator)};
-    double loss{lossAtNSigma(1.0, lossMoments) + modelSizeDifferentiator};
+    double loss{lossAtNSigma(1.0, lossMoments) +
+                this->modelSizePenalty(fixedNumberNodes, numberExtraNodes)};
     if (loss < m_BestForestTestLoss) {
         m_BestForestTestLoss = loss;
-        m_BestHyperparameters = CBoostedTreeHyperparameters{
-            m_Regularization,       m_DownsampleFactor, m_Eta,
-            m_EtaGrowthRatePerTree, maximumNumberTrees, m_FeatureBagFraction};
+        m_BestHyperparameters = CBoostedTreeHyperparameters(
+            m_Regularization, m_DownsampleFactor, m_Eta, m_EtaGrowthRatePerTree,
+            maximumNumberTrees, m_FeatureBagFraction, m_PredictionChangeCost);
     }
 }
 
+double CBoostedTreeImpl::modelSizePenalty(double fixedNumberNodes,
+                                          double numberExtraNodes) const {
+    // 0.01 * "forest number nodes" * E[GP] / "average forest number nodes" to meanLoss.
+    return 0.01 * (fixedNumberNodes + numberExtraNodes) /
+           (fixedNumberNodes + CBasicStatistics::mean(m_MeanForestSizeAccumulator)) *
+           CBasicStatistics::mean(m_MeanLossAccumulator);
+}
+
 void CBoostedTreeImpl::restoreBestHyperparameters() {
     m_Regularization = m_BestHyperparameters.regularization();
     m_DownsampleFactor = m_BestHyperparameters.downsampleFactor();
     m_Eta = m_BestHyperparameters.eta();
     m_EtaGrowthRatePerTree = m_BestHyperparameters.etaGrowthRatePerTree();
     m_MaximumNumberTrees = m_BestHyperparameters.maximumNumberTrees();
     m_FeatureBagFraction = m_BestHyperparameters.featureBagFraction();
+    m_PredictionChangeCost = m_BestHyperparameters.predictionChangeCost();
     LOG_TRACE(<< "loss* = " << m_BestForestTestLoss
```

Comment:
Maybe add the new hyperparameters to the log trace as well?

Reply:
See [5260aeb](https://github.com/elastic/ml-cpp/pull/1837/commits/5260aeb2a8e6b0d22c673a9844a1fb13aac145dd).